DSharpPlus/Entities/BaseDiscordMessageBuilder.cs

```diff
@@ -0,0 +1,296 @@
+        /// <returns></returns>
+        public abstract T AddComponents(params DiscordComponent[] components);
+
+        /// <summary>
+        /// Adds components to this message. Each call should append to a new row.
+        /// </summary>
+        /// <param name="components">Components to add.</param>
+        /// <returns></returns>
+        public abstract T AddComponents(IEnumerable<DiscordComponent> components);
+
+        /// <summary>
+        /// Adds an action row component to this message.
+        /// </summary>
+        /// <param name="components">Action row to add to this message. Should contain child components.</param>
+        /// <returns></returns>
+        public abstract T AddComponents(IEnumerable<DiscordActionRowComponent> components);
+
+        /// <summary>
+        /// Sets whether this message should play as a text-to-speech message.
+        /// </summary>
+        /// <param name="isTTS"></param>
+        /// <returns></returns>
+        public abstract T WithTTS(bool isTTS);
+
+        /// <summary>
+        /// Adds an embed to this message.
+        /// </summary>
+        /// <param name="embed">Embed to add.</param>
+        /// <returns></returns>
+        public abstract T AddEmbed(DiscordEmbed embed);
+
+        /// <summary>
+        /// Adds multiple embeds to this message.
+        /// </summary>
+        /// <param name="embeds">Collection of embeds to add.</param>
+        /// <returns></returns>
+        public abstract T AddEmbeds(IEnumerable<DiscordEmbed> embeds);
+
+        /// <summary>
+        /// Attaches a file to this message.
+        /// </summary>
+        /// <param name="fileName">Name of the file to attach.</param>
+        /// <param name="stream">Stream containing said file's contents.</param>
+        /// <param name="resetStream">Whether to reset the stream to position 0 after sending.</param>
+        /// <returns></returns>
+        public abstract T AddFile(string fileName, Stream stream, bool resetStream = false);
+
+        /// <summary>
+        /// Attaches a file to this message.
+        /// </summary>
+        /// <param name="stream">FileStream pointiong to the file to attach.</param>
+        /// <param name="resetStream">Whether to reset the stream position to 0 after sending.</param>
+        /// <returns></returns>
+        public abstract T AddFile(FileStream stream, bool resetStream = false);
+
+        /// <summary>
+        /// Attaches multiple files to this message.
+        /// </summary>
+        /// <param name="files">Dictionary of files to add, where <see cref="string"/> is a file name and <see cref="Stream"/> is a stream containing the file's contents.</param>
+        /// <param name="resetStreams">Whether to reset all stream positions to 0 after sending.</param>
+        /// <returns></returns>
+        public abstract T AddFiles(IDictionary<string, Stream> files, bool resetStreams = false);
```

Comment:
Could there be an overload that accepts `IEnumerable<DiscordMessageFile>` so we can copy files from other builders easily?

Reply:
+1